google/cloud/bigtable/_read_rows.py

```diff
@@ -0,0 +1,629 @@
+            core_exceptions.Aborted,
+        )
+
+        def on_error_fn(exc):
+            if predicate(exc):
+                self.transient_errors.append(exc)
+
+        retry = retries.AsyncRetry(
+            predicate=predicate,
+            timeout=self.operation_timeout,
+            initial=0.01,
+            multiplier=2,
+            maximum=60,
+            on_error=on_error_fn,
+            is_stream=True,
+        )
+        self._stream: AsyncGenerator[Row, None] | None = retry(
+            self._partial_retryable
+        )()
+        # contains the list of errors that were retried
+        self.transient_errors: List[Exception] = []
+
+    def __aiter__(self) -> AsyncIterator[Row]:
+        """Implements the AsyncIterable interface"""
+        return self
+
+    async def __anext__(self) -> Row:
+        """Implements the AsyncIterator interface"""
+        if self._stream is not None:
+            return await self._stream.__anext__()
+        else:
+            raise asyncio.InvalidStateError("stream is closed")
+
+    async def aclose(self):
+        """Close the stream and release resources"""
+        if self._stream is not None:
+            await self._stream.aclose()
+        self._stream = None
+        self._emitted_seen_row_key = None
+
+    async def _read_rows_retryable_attempt(
+        self,
+        gapic_fn: Callable[..., Awaitable[AsyncIterable[ReadRowsResponse]]],
+        per_request_timeout: float,
+        operation_deadline: float,
+        total_row_limit: int,
+    ) -> AsyncGenerator[Row, None]:
+        """
+        Retryable wrapper for merge_rows. This function is called each time
+        a retry is attempted.
+
+        Some fresh state is created on each retry:
+          - grpc network stream
+          - state machine to hold merge chunks received from stream
+        Some state is shared between retries:
+          - _last_emitted_row_key is used to ensure that
+            duplicate rows are not emitted
+          - request is stored and (potentially) modified on each retry
+        """
+        if self._last_emitted_row_key is not None:
+            # if this is a retry, try to trim down the request to avoid ones we've already processed
+            try:
+                self._request["rows"] = _ReadRowsOperation._revise_request_rowset(
+                    row_set=self._request.get("rows", None),
+                    last_seen_row_key=self._last_emitted_row_key,
+                )
+            except _RowSetComplete:
+                # if there are no rows left to process, we're done
+                # This is not expected to happen often, but could occur if
+                # a retry is triggered quickly after the last row is emitted
+                return
+            # revise next request's row limit based on number emitted
+            if total_row_limit:
+                new_limit = total_row_limit - self._emit_count
+                if new_limit <= 0:
```

Comment:
I think we might have to raise an error if the count goes negative. The situation would imply that there is a bug in the client or server and the results cant be trusted